✨ feat : 리뷰 기능 전체 CRUD 및 정책 로직 구현

sooyeoneo opened the pull request:
## ☀️ 작업 목적

* 일반회원, 로컬 크리에이터, 관리자가 **리뷰 및 답글을 작성·조회·수정·삭제**할 수 있는 리뷰 관리 시스템을 구축했습니다.
* 체험 완료 검증, 1회 작성 제한, soft-delete, 답글 정책 등 **리뷰 신뢰성과 데이터 무결성 유지**를 위한 정책을 구현했습니다.
* UX를 고려한 에러처리와 본문/답글 **글자수 제한(500자)** 등 정책을 API에 반영했습니다.

---

## 📄 주요 작업 내역

#### 1. Entity/Enum/Repository 설계 #122 #129

* **Review Entity**

  * **Member**, **Program**, **Reservation**과 연관관계 (`@ManyToOne`)
  * soft-delete 및 상태 관리(`ReviewStatus`, `isDeleted`, `deletedAt`)
  * 수정 이력 및 '수정됨' 상태 판단
* **ReviewReply Entity**

  * `Review`와 양방향 연관, soft-delete 지원
* **ReviewStatus Enum**: `DISPLAYED`/`DELETED` 상태 정의
* **Repository**

  * **ReviewRepository**: 상태별 조회/중복 리뷰 체크/리스트 반환
  * **ReviewReplyRepository**: 답글 존재 여부/일괄 soft-delete

#### 2. DTO/Validation #122 #123 #124 #125 #128

* **요청 DTO**

  * `ReviewRequestDto`, `ReviewReplyRequestDto`: `@NotBlank`, `@Size(500)` 등 입력 검증 어노테이션 적용
* **응답 DTO**

  * `ReviewResponseDto`: `isModified`(수정됨) 플래그, 답글 목록 포함
  * `ReviewReplyResponseDto`: 크리에이터 정보 포함
* `@Valid`: 모든 컨트롤러에 적용, Service 중복 검증 제거

#### 3. 비즈니스 로직(Service) #122 #123 #124 #126 #127 #128

* **ReviewService**

  * 리뷰 작성(체험 완료/1회 제한/30일 내/글자수 제한)
  * 리뷰 수정(본인/답글 없는 경우만/수정 플래그)
  * 리뷰 삭제(soft-delete, 답글 동시 삭제)
  * 리뷰 전체 조회(마이페이지/프로그램별/전체, soft-delete 미노출, 수정됨 표시)
* **ReviewReplyService**

  * 답글 1회 작성, 소유권 검증, 수정/단독 삭제 불가 정책
  * 답글 등록 시 리뷰 상세 반환

#### 4. REST API/Controller #122 #123 #124 #125 #126 #127

* **ReviewController**

  * 리뷰 작성/조회/수정/삭제 CRUD 엔드포인트
  * 권한 체크(`@PreAuthorize`), `CommonResponseDto` 표준 응답
* **ReviewReplyController**

  * 답글 작성 전용 API, `LOCAL_CREATOR` 권한 체크

#### 5. 정책/Validation/에러처리 #122 #128

* 리뷰/답글 본문 500자 제한
* soft-delete 및 이력관리(DB 컬럼 추가)
* 중복 리뷰, 기간 제한, 소유권 체크 등 커스텀 예외 반환
* ErrorCode 추가 및 메시지 표준화

#### 6. 인프라/Redis #122 #128

* Redis 기반 리뷰 캐싱 설정, 빈 충돌 방지(`@Qualifier`)
* Review/ReviewReply의 RedisTemplate 명시적 주입

#### 7. 테스트 및 기타 #122 #125 #130

* review\.http: API 테스트 샘플 추가
* `isModified`, `experienceEndDate` 필드 오류 수정 및 엔티티 네이밍 통일

---

## 🛠️ 추가 수정 작업 (추가 구현 예정/미구현 이슈)

* **Swagger 문서화/명세화**

  * 엔드포인트/요청·응답 명확히 문서화 필요 (현재 미완)
* **리뷰 필터링/정렬/검색/페이징**

  * 카테고리, 태그, 닉네임/본문 검색, 최신순/인기순 정렬 및 페이지네이션(#132 등) 미구현
  * `ProgramGroup` 기반 시즌제 리뷰 누적 관리(쿼리/테이블 구조) 필요
* **"이런 점이 좋아요" 태그 기능**

  * 태그는 구조상 placeholder로만 반영, 실제 태깅 및 저장/검색은 미구현(#123)
* **리뷰/답글 단위 테스트 및 정책 예외처리**

  * JUnit, MockMvc 테스트 케이스 부족(#130)
  * 각종 정책 기반 예외 및 비정상 시나리오 케이스 추가 필요
* **예약/체험 완료/기간 제한 실제 검증 로직**

  * `ReservationStatus`/`ProgramSchedule` 종료일 기반 실제 검증은 임시로 주석/로직 대체 (#123)
* **성능 최적화/DB 인덱스/쿼리 개선**

  * Soft-delete, 정렬, 대량 데이터 대응 위한 쿼리 최적화 필요

---

## 📌 참고 및 TODO

* 리뷰 1회, 30일 작성 제한, soft-delete, 답글 1회 및 단독 삭제/수정 불가 등 정책 모두 Service 계층에서 일관 적용
* 물리 삭제 금지 및 데이터 신뢰성 확보 위해 `is_deleted`, `deleted_at`, `updated_at` 관리
* 추후 `Reservation`/`ProgramSchedule` 구조 개선시 리뷰 작성/기간 검증 로직 복구 필요
* 프론트엔드와 데이터 반환 구조(pagination/리뷰+답글/태그 포함) 연동 예정

---

## 🆕 2025.07.28 추가/수정 사항

1. 소프트 딜리트 Enum 통합 및 아키텍처 일원화 #122 #127 #129
* `ReviewStatus` → `DeletedStatus`로 통합, 모든 엔티티/필드명/컬럼/메서드에 적용 

* 프로젝트 전체의 soft-delete 패턴 일관성 확보 및 중복 Enum 제거

2. 캐싱 구조 리팩토링 및 통합 #122
* `CacheKeyPrefix` 클래스 도입: Redis/Spring 캐시 프리픽스 중앙화, 인스턴스화 방지 패턴 적용 

* `ReviewCacheService` 신설: invalidate/set/get 등 캐시 로직 서비스화, 중복/직접 접근 제거, 유지보수성↑

* `RedisTemplate` 빈 주입 오류 수정 및 `@Qualifier` 명시적 생성자 주입

3. `isModified()` 플래그 정확도 개선 #122 #126
* `isModified()` 버퍼 2초 적용: `createdAt=modifiedAt`(초기화/동기화 시) 및 1ms 차이 방지, 오탐지 감소

* `modifiedAt` `null` 체크, 진짜 수정 시에만 `true`로 반환

4. 삭제 시 영속성 컨텍스트 동기화/오류 방지 #122 #127
* `ReviewService.deleteReview()`에 `EntityManager.refresh() `적용

* `@Modifying(JPQL)` 쿼리 후 `review.getReplies()` 재접근 시 삭제 데이터 노출 문제 방지

5. Repository `findByIdOrElseThrow()` 메서드 일원화 #122 #125 #130
* 서비스 로직 내 `orElseThrow` 중복 제거, Repository `default` 메서드로 통일

* 코드 가독성 및 예외 관리 일관성 강화

--- 

**리뷰 및 피드백 환영합니다! 🤗**

sooyeoneo committed:
✨ feat : 리뷰 도메인 Entity 및 Enum 구현

- Review Entity 구현
  - Member, Program, Reservation과 ManyToOne 연관관계 설정
  - ReviewStatus Enum을 통한 상태 관리
  - soft-delete 처리를 위한 deleteReview() 메서드
  - BaseEntity 상속으로 생성/수정 시간 자동 관리

- ReviewReply Entity 구현
  - Review와 ManyToOne 양방향 연관관계
  - soft-delete 지원

- ReviewStatus Enum 구현
  - DISPLAYED, DELETED 상태 정의

#122 #129

sooyeoneo committed:
✨ feat : 리뷰 Repository 인터페이스 구현

- ReviewRepository 구현
  - 상태별 조회 메서드
  - 중복 리뷰 체크
  - 페이징 처리 쿼리

- ReviewReplyRepository 구현
  - 답글 존재 여부 확인
  - 리뷰 삭제시 답글 일괄 soft-delete

#122 #127

sooyeoneo committed:
✨ feat : 리뷰 요청 DTO 구현

- ReviewRequestDto 구현
  - 리뷰 작성/수정 요청 구조

- ReviewReplyRequestDto 구현
  - 답글 작성 요청 구조

#122 #123 #124

sooyeoneo committed:
✨ feat : 리뷰 응답 DTO 구현

- ReviewResponseDto 구현
  - isModified 필드로 수정 여부 표시
  - 답글 목록 포함

- ReviewReplyResponseDto 구현
  - 크리에이터 정보 포함

#122 #125

sooyeoneo committed:
✨ feat : 리뷰 Service 비즈니스 로직 구현

- ReviewService 구현
  - 리뷰 CRUD 로직
  - 500자 제한 검증
  - Redis 캐싱 적용

- ReviewReplyService 구현
  - 답글 작성 로직
  - 프로그램 소유권 검증

#122 #123 #124 #126 #127 #128

sooyeoneo committed:
✨ feat : 리뷰 Controller REST API 구현

- ReviewController 구현
  - 리뷰 CRUD API 엔드포인트
  - Spring Security 권한 체크

- ReviewReplyController 구현
  - 답글 작성 API
  - LOCAL_CREATOR 권한 체크

#122 #123 #124 #125 #126 #127

sooyeoneo committed:
🔧 add : 리뷰 기능 ErrorCode 및 Redis 설정 추가

- ErrorCode 추가
  - 리뷰 관련 에러 코드 5종

- RedisConfig 수정
  - 리뷰 캐싱 설정

#122 #128

sooyeoneo committed:
📝 docs : 리뷰 API 테스트용 HTTP 파일 추가

- review.http 파일 추가
  - REST API 테스트 요청 샘플

#122

sooyeoneo committed:
♻️ refactor : 리뷰 조회 API 페이지네이션 제거 및 List 반환으로 변경

- ReviewController 페이지네이션 제거
  - Page<ReviewResponseDto> → List<ReviewResponseDto> 반환 타입 변경
  - Pageable 파라미터 제거 (getMyReviews, getProgramReviews, getAllReviews)
  - 불필요한 Repository 의존성 제거 (ProgramRepository, ReservationRepository)
  - ReservationService 의존성 제거

- ReviewService 페이지네이션 로직 제거
  - 조회 메서드 시그니처 변경 (Pageable 파라미터 제거)
  - Page → List 반환 타입 변경
  - stream().map().collect()를 사용한 DTO 변환 처리
  - Spring Data JPA Page 관련 import 제거

- ReviewRepository 페이지네이션 쿼리 수정
  - JPQL 쿼리에서 Pageable 파라미터 제거
  - Page<Review> → List<Review> 반환 타입 변경
  - 정렬은 JPQL ORDER BY 절로 유지 (createdAt DESC)

- Global 페이지네이션 적용 대기
  - 추후 Global 공통 페이지네이션 구현 시 재적용 예정
  - 현재는 전체 리스트 반환 방식으로 동작

#122 #125

sooyeoneo committed:
🔧 add : 리뷰 작성 시 예약 완료 상태 및 작성 기간 검증 로직 임시 수정

- validateCompletedReservation() 메서드 임시 수정
  - 예약 완료 상태 확인 로직 주석 처리
  - 모든 예약을 COMPLETED 상태로 가정하여 통과
  - 실제 ReservationStatus.COMPLETED 체크 로직은 TODO로 보존
  - 로그 메시지로 임시 로직임을 명시

- validateReviewPeriod() 메서드 임시 수정
  - Reservation 엔티티에 endDate 필드 없어 발생한 오류 해결
  - 30일 이내 작성 기간 체크 로직 주석 처리
  - 모든 리뷰 작성 기간을 유효한 것으로 가정
  - 원래 검증 로직은 주석으로 보존

- createReview() 메서드 수정
  - reservation.getEndDate() → LocalDate.now()로 임시 변경
  - TODO 주석으로 ProgramSchedule 종료일 사용 필요성 명시

- 예약 기능 완성 후 원래 로직 복구 예정
  - ProgramSchedule 또는 Program의 종료일 기준 검증
  - ReservationStatus.COMPLETED 상태 확인

#122 #123

sooyeoneo committed:
🔧 add : 리뷰 요청 DTO 검증 어노테이션 추가 및 Controller @Valid 적용

- ReviewRequestDto 검증 어노테이션 추가
  - @NotBlank: 리뷰 내용 필수값 검증
  - @Size(max = 500): 최대 500자 제한 검증
  - 에러 메시지를 한글로 명확하게 작성

- ReviewReplyRequestDto 검증 어노테이션 추가
  - @NotBlank: 답글 내용 필수값 검증
  - @Size(max = 500): 최대 500자 제한 검증
  - 에러 메시지를 한글로 명확하게 작성

- ReviewController @Valid 적용
  - createReview() 메서드에 @Valid 추가
  - updateReview() 메서드에 @Valid 추가
  - jakarta.validation.Valid import 추가

- ReviewReplyController @Valid 적용
  - createReviewReply() 메서드에 @Valid 추가
  - jakarta.validation.Valid import 추가

- 프로젝트 컨벤션 준수
  - Member, Program 도메인 검증 패턴 참고
  - 한글 에러 메시지 사용

#122 #128

sooyeoneo committed:
♻️ refactor : Service 계층의 중복된 입력값 검증 로직 제거

- validateContentLength() 메서드 제거
  - ReviewService의 validateContentLength() 메서드 삭제
  - ReviewReplyService의 validateContentLength() 메서드 삭제
  - null/빈 문자열 체크 및 500자 제한 검증 로직 제거

- 중복 검증 호출 제거
  - ReviewService.createReview()의 validateContentLength() 호출 주석 처리
  - ReviewService.updateReview()의 validateContentLength() 호출 주석 처리
  - ReviewReplyService.createReviewReply()의 validateContentLength() 호출 주석 처리

- DTO @Valid 검증으로 통합
  - @NotBlank로 빈 문자열 검증 (DTO 레벨)
  - @Size(max = 500)로 글자수 제한 검증 (DTO 레벨)
  - Controller @Valid로 자동 검증 처리

- 계층별 책임 분리
  - Controller/DTO: 입력값 형식 검증
  - Service: 비즈니스 로직 검증 (중복 리뷰, 권한 체크 등)

#122 #128

sooyeoneo committed:
🐛 fix : ReviewResponseDto의 experienceEndDate 필드 Program endDate로 변경

- ReviewResponseDto 수정
  - Reservation.getEndDate() 메서드 없어 발생한 컴파일 오류 해결
  - review.getReservation().getEndDate() → review.getProgram().getEndDate()
  - Program 엔티티의 체험 프로그램 종료일 사용

- 체험 종료일 데이터 소스 변경
  - 예약별 종료일 대신 프로그램 전체 종료일 사용
  - 추후 Reservation에 실제 체험 참여일 추가 시 재변경 예정

#122 #125

sooyeoneo committed:
🐛 fix : Review isModified() 메서드 필드명 오류 수정

- getUpdatedAt() → getModifiedAt() 변경
- BaseEntity의 실제 필드명(modifiedAt)에 맞춰 수정
- 리뷰 수정 여부 판단 로직 정상화

#122 #126

sooyeoneo committed:
♻️ refactor : Review/ReviewReply 엔티티 필드명 status → reviewStatus 변경

- Entity 필드명 변경
  - Review, ReviewReply의 status → reviewStatus
  - JPA 네이밍 규칙에 맞춰 명확한 필드명 사용

- Repository 메서드명 변경
  - findByIdAndStatus → findByIdAndReviewStatus
  - existsByReviewAndStatus → existsByReviewAndReviewStatus
  - 모든 status 관련 메서드명 일괄 변경

- Service/DTO 호출부 수정
  - Repository 메서드 호출부 전체 수정
  - ReviewResponseDto의 filter 조건 수정

#122

sooyeoneo committed:
🐛 fix : ReviewService, ReviewReplyService Redis 빈 충돌 문제 해결

- @Qualifier 어노테이션 적용
  - ReviewService 생성자에 @Qualifier("reviewRedisTemplate") 추가
  - ReviewReplyService 생성자에 @Qualifier("reviewRedisTemplate") 추가
  - RedisConfig에 정의된 특정 빈 명시적 주입

- RedisTemplate 타입 변경
  - RedisTemplate<String, Object> → RedisTemplate<String, String>
  - RedisConfig의 reviewRedisTemplate 타입과 일치

- AptitudeService 패턴 참고하여 동일하게 적용
  - 여러 RedisTemplate 빈 중복 시 @Qualifier로 구분

#122

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/review/service/ReviewService.java:
```diff
@@ -0,0 +1,204 @@
+	private final RedisTemplate<String, String> redisTemplate;
+
+	private static final String REVIEW_CACHE_KEY = "review:program:";
+	private static final long CACHE_TTL = 60; // 60분
```
이 코드는 지금 쓰이지 않는 것 같은데 나중에 TTL을 적용한 캐싱 기능을 추가할 예정이신걸까요?🧐

sooyeoneo replied:
네! 맞습니다. 🤓
현재는 캐시 무효화 기능만 구현되어 있지만, 향후 `getProgramReviews`와 같은 조회 API에 Redis 캐싱 기능을 추가하면서 
이 `CACHE_TTL` 변수를 활용하여 만료 시간을 설정할 계획이라 미리 변수를 정의하였습니다.
감사합니다!

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/review/entity/Review.java:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	public boolean isModified() {
+		return getModifiedAt() != null && getModifiedAt().isAfter(getCreatedAt());
```
현재 `isModified()` 메서드에서 DB 동기화나 Auditing 초기화로 인해 실제 수정하지 않았는데도 `true`로 인식될 수 있을 것 같아요.

- `modifiedAt`이 `createdAt`과 거의 같은 시간으로 찍히는 경우 (1ms 차이도 `true`)
- 엔티티 생성 시 Auditing이 `modifiedAt`을 초기화하는 경우

이를 방지하려면 버퍼 시간을 주어 진짜 수정만 감지하도록 하면 좋을 것 같아요🥹

```java
public boolean isModified() {
    if (getModifiedAt() == null) return false;

    return !getModifiedAt().isEqual(getCreatedAt()) &&
           getModifiedAt().isAfter(getCreatedAt().plusSeconds(2));
}
```
이렇게 하면 DB 동기화나 초기화로 인한 불필요한 감지를 막고, 실제 수정된 경우만 true로 반환할 수 있을 것 같습니다!☺️

- `isEqual` : 1ms 차이도 수정으로 잘못 인식될 수 있음
- `isAfter` : 생성 직후 `modifiedAt` 초기화 때문에 의도치 않게 `true`가 될 수 있음
- `plusSeconds(2)` : 진짜 수정된 경우만 감지하고, DB 동기화 차이는 무시

sooyeoneo replied:
`isModified()` 메서드의 문제점을 명확히 알려주셔서 감사합니다. ☺️
말씀해주신 대로 Auditing 초기화나 DB 동기화로 인해 의도치 않게 `true`로 인식될 수 있는 문제점이 있네요..

제안해주신 버퍼 시간을 두는 방식으로 코드를 수정했어요!
```
getModifiedAt().isAfter(getCreatedAt().plusSeconds(2)) 
```
조건을 사용하여 생성 후 최소 2초 이후에 `modifiedAt`이 찍힌 경우에만 실제 수정으로 감지하도록 변경 완료했습니다.
이렇게 하면, 불필요한 `true` 반환을 막고 실제 사용자 수정만 정확하게 판별할 수 있을 것 같네요.

다시 한번 꼼꼼한 리뷰 감사드립니다! 💗

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/review/service/ReviewReplyService.java:
```diff
@@ -0,0 +1,81 @@
+	}
+
+
+	private void invalidateReviewCache(Long programId) {
```
현재는 프로그램 단일 캐시 키(`review:program:{programId}`)만 삭제하지만, 추후 페이지네이션이나 정렬 옵션별로 캐시를 따로 둘 경우를 생각하여 SCAN 기반으로 패턴을 매칭해 연관된 캐시를 한 번에 삭제하는 방식으로 확장하는 것은 어떨까요?🤔

```java
private void invalidateReviewCache(Long programId) {
    String pattern = REVIEW_CACHE_KEY + programId + "*";
    redisTemplate.keys(pattern).forEach(redisTemplate::delete);
}
```

sooyeoneo replied:
유익한 코멘트 감사합니다. 말씀해주신 부분을 반영하여 다음과 같이 개선했습니다. 😎🖤

1. **캐시 키 상수 통합**: `CacheKeyPrefix` 클래스로 캐시 키를 한 곳에서 관리하도록 변경

2. **캐시 서비스 분리**: `ReviewCacheService`를 새로 만들어 캐싱 로직을 중앙화 
  → ReviewService와 ReviewReplyService의 중복 코드 제거

3. **패턴 매칭 삭제 지원**: `invalidateReviewCacheByPattern()` 메서드를 구현하여 
향후 페이지네이션이나 정렬 옵션별 캐시 확장에 대비

```java
// ReviewCacheService.java
public void invalidateReviewCacheByPattern(Long programId) {
    String pattern = CacheKeyPrefix.REVIEW_PROGRAM + programId + "*";
    Set<String> keys = redisTemplate.keys(pattern);
    
    if (keys != null && !keys.isEmpty()) {
        redisTemplate.delete(keys);
    }
}

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/review/service/ReviewReplyService.java:
```diff
@@ -0,0 +1,81 @@
+	private final ReviewReplyRepository reviewReplyRepository;
+	private final RedisTemplate<String, String> redisTemplate;
+
+	private static final String REVIEW_CACHE_KEY = "review:program:";
```
이후 다른 도메인에서도 캐시 키 프리픽스를 쓸 가능성이 있고, `private static final String REVIEW_CACHE_KEY = "review:program:";`이 `ReviewService, ReviewReplyService` 두 곳에서 중복 선언되고 있어 공통 상수로 빼서 관리하는 것이 좋을 것 같습니다!😎

```java

public final class CacheKeyPrefix {

    public static final String REVIEW_PROGRAM = "review:program:"; 

    private CacheKeyPrefix() {
    }
}
```
이렇게 하게 되면 프리픽스 변경 시 한 곳만 수정하면 되고, 다른 도메인 캐시 키도 한번에 볼 수 있어 키 충돌 방지에 도움이 될 것 같아요!☺️

sooyeoneo replied:
자세한 코멘트 감사합니다! 😊

말씀해주신 대로 `CacheKeyPrefix`라는 별도의 공통 상수 클래스를 `global/util` 패키지 경로에 생성하여 
(`constant`라는 패키지에 추가하는 것이 관례인 것 같지만, 프로젝트 규모가 훨씬 커져서 Redis 캐시 키 외의 다른 공통 상수들이 많이 생기면, 그때 `global/constant` 패키지를 분리하겠습니다!)

`REVIEW_PROGRAM` 캐시 키를 한 곳에서 관리하도록 수정했습니다. 
덕분에 코드의 응집성을 높이고 향후 캐시 키 관리 및 충돌 방지에 큰 도움이 될 것 같습니다. 

```
public final class CacheKeyPrefix {

	// 리뷰 관련 캐시 키 프리픽스
	public static final String REVIEW_PROGRAM_BASE = "review:program:"; // 실제 Redis 키 패턴 매칭용
	public static final String PROGRAM_REVIEWS_CACHE_NAME = "programReviews"; // Spring Cache @Cacheable/@CacheEvict value 이름

	private CacheKeyPrefix() {
		// 인스턴스화를 방지하기 위한 private 생성자
	}
}

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/review/service/ReviewService.java:
```diff
@@ -0,0 +1,204 @@
+
+		// Soft delete
+		review.deleteReview();
+
```
`softDeleteByReview()`는 JPQL UPDATE라서 DB에는 즉시 반영되지만, 이미 로드된 review.getReplies() 컬렉션은 영속성 컨텍스트에 그대로 남아있어 같은 트랜잭션 안에서 다시 접근하면 삭제된 답글이 보일 수 있습니다.

이를 방지하기 위해 entityManager.refresh(review)로 엔티티를 DB 상태와 동기화하는 게 좋을 것 같아요☺️

```java
    @Transactional
    public void deleteReview(Long reviewId, Member member) {
        log.info("리뷰 삭제 - reviewId: {}, memberId: {}", reviewId, member.getId());

        Review review = reviewRepository.findByIdAndReviewStatusOrThrow(reviewId, ReviewStatus.DISPLAYED);

        // 본인 확인
        if (!review.getMember().getId().equals(member.getId())) {
            throw new CustomException(ErrorCode.UNAUTHORIZED);
        }

        // Soft delete
        review.deleteReview();

        // 답글도 함께 soft delete
        reviewReplyRepository.softDeleteByReview(review);

        // 영속성 컨텍스트를 최신 DB 상태로 동기화 -> 추가
        entityManager.refresh(review);

        // 캐시 무효화 (Redis)
        invalidateReviewCache(review.getProgram().getId());
    }
```

sooyeoneo replied:
정확한 지적 감사합니다! 🙏

`@Modifying` JPQL 쿼리가 영속성 컨텍스트를 거치지 않아 DB와의 불일치가 발생할 수 있다는 점을 놓치고 있었네요. 

제안해주신 `entityManager.refresh(review)`를 추가하여 영속성 컨텍스트를 DB 상태와 동기화하도록 수정했습니다. 현재는 `deleteReview()` 메서드가 void 반환이라 당장은 문제가 없지만, 향후 같은 트랜잭션 내에서 review 엔티티를 재사용할 경우를 대비한 안전한 코드가 될 것 같습니다.

JPA 영속성 컨텍스트의 동작 방식에 대한 심도있는 리뷰 감사드립니다! 😊

j-hann commented on src/main/java/com/salayo/locallifebackend/domain/review/enums/ReviewStatus.java:
```diff
@@ -0,0 +1,7 @@
+package com.salayo.locallifebackend.domain.review.enums;
+
+public enum ReviewStatus {
```
global쪽 enum에 똑같은 목적으로 사용하는 DeletedStatus가 있는데 enum 상수도 똑같아서 말씀드립니다! 혹시 다른 이유가 없다면 DeletedStatus enum으로 사용하는 게 좋을 것 같다는 의견입니다~!😊

sooyeoneo replied:
코멘트 감사합니다! 😊 확인해보니 정말 `DeletedStatus`와 `ReviewStatus`가 동일하네요. 

말씀하신 대로 `DeletedStatus`를 사용하는 것이 맞다고 생각해요. 
프로젝트 전체에서 soft-delete 패턴을 일관되게 사용할 수 있고, 중복 코드도 제거할 수 있겠네요!

`Review.java`, `ReviewReply.java`의 `ReviewStatus`를 `DeletedStatus`로 모두 교체하였습니다! ❣️

j-hann commented on src/main/java/com/salayo/locallifebackend/domain/review/service/ReviewService.java:
```diff
@@ -0,0 +1,204 @@
+			.orElseThrow(() -> new CustomException(ErrorCode.PROGRAM_NOT_FOUND));
+		
+		Reservation reservation = reservationRepository.findById(reservationId)
+			.orElseThrow(() -> new CustomException(ErrorCode.RESERVATION_NOT_FOUND));
```
수연님! `Program`이랑 `Reservation` id 기반 조회 많이 사용할 것 같아서 `ProgramRepository`랑 `ReservationRepository`에 

```
default Program findByIdOrElseThrow(long programId){
		return findById(programId).orElseThrow(() -> new CustomException(ErrorCode.PROGRAM_NOT_FOUND));
}
```
	

```
default Reservation findByIdOrElseThrow(long reservationId){
		return findById(reservationId).orElseThrow(() -> new CustomException(ErrorCode.RESERVATION_NOT_FOUND));
}
```

이렇게 공통 메서드 추가해놨습니다! `ErrorCode` 는 수연님이 작성해주신거 기준으로 반영해서 넣었습니다~!
참고하셔서 서비스 단에서 예외 처리 해주신거 `findByIdOrElseThrow()`로 수정해주시면 좋을 것 같아요오🤗

sooyeoneo replied:
네! `ProgramRepository`랑 `ReservationRepository`에서 확인해서 추가했습니다~ 🙌

`findByIdOrElseThrow()` 메서드 
- `createReview()`: Program, Reservation 조회 부분
- `getProgramReviews()`: Program 조회 부분

감사합니다! 😊

sooyeoneo committed:
🐛 fix  : Review isModified() 메서드 시간 버퍼 적용으로 정확도 개선

- isModified() 메서드 개선
  - modifiedAt null 체크 추가
  - createdAt과 정확히 같거나 이전인 경우 false 반환
  - 2초 버퍼 시간 적용하여 실제 수정만 감지

- 기존 문제점 해결
  - 1ms 차이도 수정으로 인식하는 과도한 민감도 해결
  - Auditing 초기화로 인한 오탐지 방지
  - DB 동기화 지연으로 인한 미세한 시간차 무시

- 수정 감지 정확도 향상
  - 생성 후 2초 이내는 초기화/동기화로 간주
  - 실제 사용자 수정 행위만 true로 반환

#122 #126

sooyeoneo committed:
♻️ refactor : 캐시 키 프리픽스 상수화를 위한 CacheKeyPrefix 클래스 추가

- CacheKeyPrefix 클래스 생성 (global/util)
  - REVIEW_PROGRAM: Redis 키 패턴 매칭용 프리픽스
  - PROGRAM_REVIEW_CACHE: Spring Cache 어노테이션용 캐시 이름
  - private 생성자로 인스턴스화 방지

- 캐시 키 중앙 관리 체계 구축
  - ReviewService, ReviewReplyService 중복 상수 제거 준비
  - 향후 다른 도메인 캐시 키 추가 시 통합 관리 가능

#122

sooyeoneo committed:
♻️ refactor : 리뷰 캐싱 로직 중복 제거 및 캐시 서비스 분리

- CacheKeyPrefix 클래스로 캐시 키 상수 통합
  - REVIEW_PROGRAM 키로 통일
  - 불필요한 PROGRAM_REVIEW_CACHE 상수 제거

- ReviewCacheService 생성하여 캐싱 로직 중앙화
  - invalidateReviewCache(): 단일 키 삭제
  - invalidateReviewCacheByPattern(): 패턴 매칭 삭제
  - getReviewCache(), setReviewCache(): 캐시 조회/저장

- ReviewService, ReviewReplyService 리팩토링
  - 중복된 invalidateReviewCache() 메서드 제거
  - RedisTemplate 직접 의존성을 ReviewCacheService로 대체
  - 캐시 무효화 로직을 서비스 호출로 통일

캐시 관련 로직이 한 곳에서 관리되어 유지보수성과 확장성이 향상됨

#122

sooyeoneo committed:
💡comment : 주석 스타일 변경

sooyeoneo committed:
🔥 remove : ReviewStatus enum 삭제

  - DeletedStatus와 완전히 동일한 상수값(DISPLAYED, DELETED)으로 중복 제거

 #122 #127 #129

sooyeoneo committed:
♻️ refactor : ReviewStatus를 DeletedStatus로 통일하여 중복 제거

- Review, ReviewReply 엔티티의 reviewStatus를 deletedStatus로 변경
  - 컬럼명도 review_status → deleted_status, reply_status → deleted_status로 변경
  - 프로젝트 전체의 soft delete 패턴 일관성 확보

- Repository 메서드명 및 파라미터 타입 변경
  - findByIdAndReviewStatus → findByIdAndDeletedStatus
  - existsByReviewAndReviewStatus → existsByReviewAndDeletedStatus 등

- Service, DTO에서 DeletedStatus 사용하도록 수정
  - ReviewStatus.DISPLAYED → DeletedStatus.DISPLAYED
  - import 문 정리

#122 #127 #129

sooyeoneo committed:
🐛 fix : 리뷰 삭제 시 영속성 컨텍스트 동기화 문제 해결

- ReviewService.deleteReview()에 EntityManager.refresh() 추가
- JPQL UPDATE 쿼리로 답글 삭제 시 영속성 컨텍스트가 DB 상태를 반영하지 못하는 문제 수정
- 같은 트랜잭션 내에서 review.getReplies() 재접근 시 삭제된 답글이 보이는 현상 방지

softDeleteByReview()는 @Modifying 쿼리로 DB에 직접 반영되지만,
영속성 컨텍스트의 Review 엔티티는 이를 알지 못하므로 refresh() 필요

#122 #127

sooyeoneo committed:
🐛 fix : ReviewCacheService 빈 주입 오류 수정 및 불필요한 메서드 제거

- @RequiredArgsConstructor와 @Qualifier 호환성 문제 해결
  - 명시적 생성자로 변경하여 @Qualifier를 파라미터에 직접 적용
  - 5개의 RedisTemplate 빈 중 reviewRedisTemplate 정확히 주입

- 사용하지 않는 invalidateReviewCache() 메서드 제거
  - 현재 모든 서비스에서 invalidateReviewCacheByPattern()만 사용
  - YAGNI 원칙에 따라 불필요한 코드 정리

 #122

sooyeoneo committed:
Merge branch 'develop' into feature/review

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
♻️ refactor : Repository의 findByIdOrElseThrow() 메서드 활용

- ReviewService에서 findById().orElseThrow() 패턴을 findByIdOrElseThrow()로 교체
  - programRepository.findByIdOrElseThrow()
  - reservationRepository.findByIdOrElseThrow()

- 중복된 예외 처리 로직을 Repository default 메서드로 통합하여 코드 간결화
- 동일한 ErrorCode 사용으로 일관성 있는 예외 처리 보장

#122 #125 #130
